### ISSUE TITLE
Including video.js in a project produces a security advisory on GitHub

### ISSUE THREAD

**Issue by dj-documentation** [videojs/video.js#7389]
## Description
For the past 5 months or so video.js has included dependencies with security vulnerabilities, causing `npm audit`, `yarn audit`, and GitHub to display warnings. Though the security vulnerabilities may not be serious leaving video.js unpatched is training users to ignore security alerts. What needs to be done to get the dependencies updated? 

![image](https://user-images.githubusercontent.com/88207951/130303590-331c2685-ac6e-4dbc-a51b-157bba814c4e.png)

About alerts for vulnerable dependencies
https://docs.github.com/en/code-security/supply-chain-security/managing-vulnerabilities-in-your-projects-dependencies/about-alerts-for-vulnerable-dependencies

## Steps to reproduce
1. Add the latest commit of video.js to package.json with `"video.js": "git://github.com/videojs/video.js"`
2. A security advisory is shown at the top of your repository's main page
3. `npm audit` and `yarn audit` also show the advisory

### Expected
`npm audit` and `yarn audit` should be clean

### Actual
`yarn audit` shows advisories
```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Misinterpretation of malicious XML input                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.7.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ video.js                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ video.js > @videojs/http-streaming > mpd-parser > xmldom     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1769                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Misinterpretation of malicious XML input                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.7.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ video.js                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ video.js > @videojs/http-streaming > video.js >              │
│               │ @videojs/http-streaming > mpd-parser > xmldom                │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1769                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Misinterpretation of malicious XML input                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.7.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ video.js                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ video.js > @videojs/http-streaming > video.js > mpd-parser > │
│               │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1769                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Misinterpretation of malicious XML input                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.7.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ video.js                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ video.js > mpd-parser > xmldom                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1769                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

**Comment by dj-documentation** [comment]
mpd-parser is the package that includes the xmldom

https://github.com/videojs/mpd-parser/blob/fd85fd65ea494d989bf71f12a84b96c3395914b9/package.json#L62

**Comment by gkatsev** [comment]
It's worth noting that this won't be an issue if you are a web-facing project. xmldom is *only* used by mpd-parser when it's run via the command line.

**Comment by gkatsev** [comment]
However, we'll take a look next week.

**Comment by dj-documentation** [comment]
This is still an issue. The latest pre-release (version 7.15.4) causes an alert.

**Comment by gkatsev** [comment]
@dj-documentation I think it might be an npm caching issue. If you still have it showing a warning, can you run `npm ls mpd-parser` and post the output?
To me, I was still seeing an older version of Video.js in the tree but after uninstalling and re-installing some things, I no longer see an older version of Video.js in the tree and the audit doesn't show any issues anymore.
working:
```sh
$ npm ls mpd-parser
videojs-npm-test@1.0.0 ~/videojs-npm-test
└─┬ video.js@7.15.4
  ├─┬ @videojs/http-streaming@2.10.2
  │ └── mpd-parser@0.19.0  deduped
  └── mpd-parser@0.19.0
```
broken
```sh
$ npm ls mpd-parser
videojs-npm-test@1.0.0 /Users/gkatsevman/p/videojs-npm-test
└─┬ @videojs/http-streaming@2.10.3
  ├── mpd-parser@0.19.0
  └─┬ video.js@7.14.3
    ├─┬ @videojs/http-streaming@2.9.2
    │ └── mpd-parser@0.17.0  deduped
    └── mpd-parser@0.17.0
```
I would expect `npm cache clear` and reinstall should kick it into gear. Or potentially `npm dedupe`.

**Comment by dj-documentation** [comment]
We're unable to run `npm cache clear --force` on [dependabot](https://dependabot.com/) because it's a service bot run by GitHub. Is there something else we can try?
 
The advisory seems to be caused by either mpd-parser or http-streaming.

```
$ npm ls video.js
test@1.0.0 ~/test
└─┬ video.js@7.15.4
  └─┬ @videojs/http-streaming@2.10.2
    └── video.js@7.15.4 deduped
```
```
$ npm ls mpd-parser
test@1.0.0 ~/test
└─┬ video.js@7.15.4
  ├─┬ @videojs/http-streaming@2.10.2
  │ └── mpd-parser@0.19.0
  └── mpd-parser@0.19.0
```
```
$ npm ls @videojs/http-streaming
test@1.0.0 ~/test
└─┬ video.js@7.15.4
  └── @videojs/http-streaming@2.10.2
```

**Comment by dj-documentation** [comment]
Output from `npm view`:

```
$ npm view video.js dependencies
{
  '@babel/runtime': '^7.12.5',
  '@videojs/http-streaming': '2.9.2',
  '@videojs/vhs-utils': '^3.0.2',
  '@videojs/xhr': '2.5.1',
  'aes-decrypter': '3.1.2',
  global: '^4.4.0',
  keycode: '^2.2.0',
  'm3u8-parser': '4.7.0',
  'mpd-parser': '0.17.0',
  'mux.js': '5.12.2',
  'safe-json-parse': '4.0.0',
  'videojs-font': '3.2.0',
  'videojs-vtt.js': '^0.15.3'
}
```
```
$ npm view mpd-parser dependencies
{
  '@babel/runtime': '^7.12.5',
  '@videojs/vhs-utils': '^3.0.2',
  '@xmldom/xmldom': '^0.7.2',
  global: '^4.4.0'
}
```
```
$ npm view @videojs/http-streaming dependencies
{
  '@babel/runtime': '^7.12.5',
  '@videojs/vhs-utils': '3.0.3',
  'aes-decrypter': '3.1.2',
  global: '^4.4.0',
  'm3u8-parser': '4.7.0',
  'mpd-parser': '0.19.0',
  'mux.js': '5.13.0',
  'video.js': '^6 || ^7'
}
```

**Comment by dj-documentation** [comment]
yarn.lock contains:
<details>
<summary>yarn.lock</summary>

```
"@videojs/http-streaming@2.10.2":
  version "2.10.2"
  resolved "https://registry.yarnpkg.com/@videojs/http-streaming/-/http-streaming-2.10.2.tgz#02e6fcfa74f7850b5f9eb40a8e5c85c9d7d33eaf"
  integrity sha512-JTAlAUHzj0sTsge2WBh4DWKM2I5BDFEZYOvzxmsK/ySILmI0GRyjAHx9uid68ZECQ2qbEAIRmZW5lWp0R5PeNA==
  dependencies:
    "@babel/runtime" "^7.12.5"
    "@videojs/vhs-utils" "3.0.3"
    aes-decrypter "3.1.2"
    global "^4.4.0"
    m3u8-parser "4.7.0"
    mpd-parser "0.19.0"
    mux.js "5.13.0"
    video.js "^6 || ^7"

"@videojs/http-streaming@2.9.2":
  version "2.9.2"
  resolved "https://registry.yarnpkg.com/@videojs/http-streaming/-/http-streaming-2.9.2.tgz#47d33bb02bd9c1287200398b1e85d213dee814d0"
  integrity sha512-2ZsxJn4/nZZ6k6jIhic2l9ynGmKwprtuI5b3+M6JgqOSLvQQ/ah+heVs/0g2Ze7qJxodqR+aSY948JwJIz1gCw==
  dependencies:
    "@babel/runtime" "^7.12.5"
    "@videojs/vhs-utils" "^3.0.2"
    aes-decrypter "3.1.2"
    global "^4.4.0"
    m3u8-parser "4.7.0"
    mpd-parser "0.17.0"
    mux.js "5.12.2"
    video.js "^6 || ^7"

"@videojs/vhs-utils@3.0.3", "@videojs/vhs-utils@^3.0.0", "@videojs/vhs-utils@^3.0.2", "@videojs/vhs-utils@^3.0.3":
  version "3.0.3"
  resolved "https://registry.yarnpkg.com/@videojs/vhs-utils/-/vhs-utils-3.0.3.tgz#708bc50742e9481712039695299b32da6582ef92"
  integrity sha512-bU7daxDHhzcTDbmty1cXjzsTYvx2cBGbA8hG5H2Gvpuk4sdfuvkZtMCwtCqL59p6dsleMPspyaNS+7tWXx2Y0A==
  dependencies:
    "@babel/runtime" "^7.12.5"
    global "^4.4.0"
    url-toolkit "^2.2.1"

"@videojs/xhr@2.5.1":
  version "2.5.1"
  resolved "https://registry.yarnpkg.com/@videojs/xhr/-/xhr-2.5.1.tgz#26bc5a79dbb3b03bfb13742c6ce559f89e90719e"
  integrity sha512-wV9nGESHseSK+S9ePEru2+OJZ1jq/ZbbzniGQ4weAmTIepuBMSYPx5zrxxQA0E786T5ykpO8ts+LayV+3/oI2w==
  dependencies:
    "@babel/runtime" "^7.5.5"
    global "~4.4.0"
    is-function "^1.0.1"

"@videojs/xhr@2.6.0":
  version "2.6.0"
  resolved "https://registry.yarnpkg.com/@videojs/xhr/-/xhr-2.6.0.tgz#cd897e0ad54faf497961bcce3fa16dc15a26bb80"
  integrity sha512-7J361GiN1tXpm+gd0xz2QWr3xNWBE+rytvo8J3KuggFaLg+U37gZQ2BuPLcnkfGffy2e+ozY70RHC8jt7zjA6Q==
  dependencies:
    "@babel/runtime" "^7.5.5"
    global "~4.4.0"
    is-function "^1.0.1"

"@xmldom/xmldom@^0.7.2":
  version "0.7.4"
  resolved "https://registry.yarnpkg.com/@xmldom/xmldom/-/xmldom-0.7.4.tgz#93b2f9486c88b6464e97f76c9ab49b0a548fbe57"
  integrity sha512-wdxC79cvO7PjSM34jATd/RYZuYWQ8y/R7MidZl1NYYlbpFn1+spfjkiR3ZsJfcaTs2IyslBN7VwBBJwrYKM+zw==

"video.js@^6 || ^7":
  version "7.14.3"
  resolved "https://registry.yarnpkg.com/video.js/-/video.js-7.14.3.tgz#0b612c09a0a81ef9bce65c710e73291cb06dc32c"
  integrity sha512-6avCdSIfn5ss5NOgoQfY/xEfPNcz9DXSw+ZN80NwPguCdRd4VL4y40b/d7osYJwyCdF+YkvhqAW7dw4s0vBigg==
  dependencies:
    "@babel/runtime" "^7.12.5"
    "@videojs/http-streaming" "2.9.2"
    "@videojs/vhs-utils" "^3.0.2"
    "@videojs/xhr" "2.5.1"
    aes-decrypter "3.1.2"
    global "^4.4.0"
    keycode "^2.2.0"
    m3u8-parser "4.7.0"
    mpd-parser "0.17.0"
    mux.js "5.12.2"
    safe-json-parse "4.0.0"
    videojs-font "3.2.0"
    videojs-vtt.js "^0.15.3"

video.js@^7.15.4:
  version "7.15.4"
  resolved "https://registry.yarnpkg.com/video.js/-/video.js-7.15.4.tgz#0f96ef138035138cb30bf00a989b6174f0d16bac"
  integrity sha512-hghxkgptLUvfkpktB4wxcIVF3VpY/hVsMkrjHSv0jpj1bW9Jplzdt8IgpTm9YhlB1KYAp07syVQeZcBFUBwhkw==
  dependencies:
    "@babel/runtime" "^7.12.5"
    "@videojs/http-streaming" "2.10.2"
    "@videojs/vhs-utils" "^3.0.3"
    "@videojs/xhr" "2.6.0"
    aes-decrypter "3.1.2"
    global "^4.4.0"
    keycode "^2.2.0"
    m3u8-parser "4.7.0"
    mpd-parser "0.19.0"
    mux.js "5.13.0"
    safe-json-parse "4.0.0"
    videojs-font "3.2.0"
    videojs-vtt.js "^0.15.3"

videojs-font@3.2.0:
  version "3.2.0"
  resolved "https://registry.yarnpkg.com/videojs-font/-/videojs-font-3.2.0.tgz#212c9d3f4e4ec3fa7345167d64316add35e92232"
  integrity sha512-g8vHMKK2/JGorSfqAZQUmYYNnXmfec4MLhwtEFS+mMs2IDY398GLysy6BH6K+aS1KMNu/xWZ8Sue/X/mdQPliA==

videojs-vtt.js@^0.15.3:
  version "0.15.3"
  resolved "https://registry.yarnpkg.com/videojs-vtt.js/-/videojs-vtt.js-0.15.3.tgz#84260393b79487fcf195d9372f812d7fab83a993"
  integrity sha512-5FvVsICuMRx6Hd7H/Y9s9GDeEtYcXQWzGMS+sl4UX3t/zoHp3y+isSfIPRochnTH7h+Bh1ILyC639xy9Z6kPag==
  dependencies:
    global "^4.3.1"

xmldom@^0.5.0:
  version "0.5.0"
  resolved "https://registry.yarnpkg.com/xmldom/-/xmldom-0.5.0.tgz#193cb96b84aa3486127ea6272c4596354cb4962e"
  integrity sha512-Foaj5FXVzgn7xFzsKeNIde9g6aFBxTPi37iwsno8QvApmtg7KYrr+OPyRHcJF7dud2a5nGRBXK3n0dL62Gf7PA==
```

</details>

**Comment by dj-documentation** [comment]
Output from `yarn audit`:

```
$ yarn audit
yarn audit v1.22.11
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Misinterpretation of malicious XML input                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.7.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ video.js                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ video.js > @videojs/http-streaming > video.js >              │
│               │ @videojs/http-streaming > mpd-parser > xmldom                │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1769                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Misinterpretation of malicious XML input                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.7.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ video.js                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ video.js > @videojs/http-streaming > video.js > mpd-parser > │
│               │ xmldom                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1769                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
2 vulnerabilities found - Packages audited: 873
Severity: 2 Moderate
✨  Done in 0.61s.
```

**Comment by dj-documentation** [comment]
Thanks for looking into this.

**Comment by gkatsev** [comment]
It looks like your yarn lock file contains multiple versions of Video.js, one of which is 7.14.3, which is likely why you're still seeing issues.
I've found that yarn does not do a good job of deduping things. I would recommend running https://github.com/atlassian/yarn-deduplicate to fix it up.

**Comment by gkatsev** [comment]
VHS depends on a loose version of Video.js (https://github.com/videojs/http-streaming/blob/a8d3c1a62f31323cc9fdbf5846529f5626af6906/package.json#L65) and relies on the package manager client to properly dedupe the versions.
If the yarn.lock doesn't have old versions of Video.js in it, github or other audit stuff won't show a warning.

I just marked 7.15.4 as latest. From what I found when tracking down various yarn related issues, it has less issues with deduping Video.js when you're updating to the latest version. If you re-install video.js 7.15.4, it should hopefully fix it.
